diff --git a/init.go b/init.go
index 31f3604..7f0c1d5 100644
--- a/init.go
+++ b/init.go
@@ -222,29 +222,34 @@ func launch(cmd string, args []string, env []string) (pd int, err error) {
 // consider using https://github.com/hashicorp/go-reap
 // which is perfectly suited for this
-func reapUntil(pid int) syscall.Signal {
+func reapUntil(pid int) int {
 	for {
 		status := new(syscall.WaitStatus)
-		wpid, err := syscall.Wait4(-1, status, syscall.WNOHANG, nil)
-		switch err {
-		case nil:
-			// the child we were waiting for died, return and pass the exit status
-			if wpid == pid {
-				if status.Exited() {
-					if status.Signal() != 0 {
-						warn("child exited with error")
-					}
-					return status.Signal()
-				}
-				warn("child exited by signal")
-				return 128 + status.Signal()
+		wpid, err := syscall.Wait4(-1, status, 0, nil)
+		if err != nil {
+			switch {
+			case errors.Is(err, syscall.EINTR):
+				continue
+			case errors.Is(err, syscall.ECHILD):
+				return 0
+			default:
+				return 128
 			}
+		}
 
-		case syscall.ECHILD:
-			// no more children
-			return 0
-		default:
-			// an error we did not expect
-			return 128
+		if wpid != pid {
+			continue
+		}
+
+		if status.Exited() {
+			exitStatus := status.ExitStatus()
+			if exitStatus != 0 {
+				warn("child exited with error")
+			}
+			return exitStatus
+		}
+		if status.Signaled() {
+			warn("child exited by signal")
+			return 128 + int(status.Signal())
 		}
 	}
 }
@@ -376,9 +381,9 @@ func main() {
 
 	warn("launched cmd=" + cmd[0])
 
-	signal := reapUntil(pid)
+	status := reapUntil(pid)
 
-	warn("reaped all children, returned with signal=" + signal.String())
+	warn(fmt.Sprintf("reaped all children, returned with status=%d", status))
 
 	_ = unix.Reboot(RB_AUTOBOOT)
 }
