diff --git a/vhost-device-vsock/src/vhu_vsock_thread.rs b/vhost-device-vsock/src/vhu_vsock_thread.rs
index 2a05e5692..a08a02027 100644
--- a/vhost-device-vsock/src/vhu_vsock_thread.rs
+++ b/vhost-device-vsock/src/vhu_vsock_thread.rs
@@ -356,13 +356,6 @@ impl VhostUserVsockThread {
                                     local_port,
                                     peer_port,
                                 );
-                                if let Err(err) = Self::epoll_register(
-                                    self.get_epoll_fd(),
-                                    stream_raw_fd,
-                                    epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
-                                ) {
-                                    warn!("Failed to register with epoll: {err:?}");
-                                }
                             }
                             Err(err) => {
                                 warn!("Unable to accept new local connection: {err:?}");
@@ -420,13 +413,8 @@ impl VhostUserVsockThread {
 
                         self.add_new_connection_from_host(fd, stream, local_port, peer_port);
 
-                        // Re-register the fd to listen for EPOLLIN and EPOLLOUT events
-                        Self::epoll_modify(
-                            self.get_epoll_fd(),
-                            fd,
-                            epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
-                        )
-                        .unwrap();
+                        // Wait until the guest accepts before forwarding local data.
+                        Self::epoll_unregister(self.get_epoll_fd(), fd).unwrap();
                     }
                 }
             } else {
diff --git a/vhost-device-vsock/src/vsock_conn.rs b/vhost-device-vsock/src/vsock_conn.rs
index aee4fc124..f83c77dcf 100644
--- a/vhost-device-vsock/src/vsock_conn.rs
+++ b/vhost-device-vsock/src/vsock_conn.rs
@@ -238,6 +238,21 @@ impl<S: AsRawFd + ReadVolatile + Write + WriteVolatile + IsHybridVsock> VsockCon
                     self.stream.write_all(response.as_bytes()).unwrap();
                 }
                 self.connect = true;
+                if VhostUserVsockThread::epoll_register(
+                    self.epoll_fd,
+                    self.stream.as_raw_fd(),
+                    epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+                )
+                .is_err()
+                {
+                    if let Err(e) = VhostUserVsockThread::epoll_modify(
+                        self.epoll_fd,
+                        self.stream.as_raw_fd(),
+                        epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+                    ) {
+                        error!("epoll_register failed: {e:?}, but proceed further.");
+                    }
+                }
             }
             VSOCK_OP_RW => {
                 // Data has to be written to the host-side stream
